File: lms/lms/report/lms_sales_pipeline/lms_sales_pipeline.js
frappe.query_reports["LMS Sales Pipeline"] = {
	filters: [
		{
			fieldname: "contract_status",
			label: "Status",
			fieldtype: "Select",
			options: "\nOngoing\nCompleted"
		},
		{
			fieldname: "customer",
			label: "Customer",
			fieldtype: "Link",
			options: "Customer"
		},
		{
			fieldname: "from_date",
			label: "From Date",
			fieldtype: "Date"
		},
		{
			fieldname: "to_date",
			label: "To Date",
			fieldtype: "Date",
			default: frappe.datetime.get_today()
		}
	],
	formatter: function (value, row, column, data, default_formatter) {
		let formatted = default_formatter(value, row, column, data);

		if (!data) {
			return formatted;
		}

		if (column.fieldname === "contract_status") {
			const styles = {
				"Ongoing": "background:#fff4e6;color:#9c5c00;",
				"Completed": "background:#e6fcf0;color:#1f7a3f;",
				"Terminated": "background:#fff3bf;color:#8f5a00;",
				"Cancelled": "background:#fff5f5;color:#c92a2a;",
				"Draft": "background:#f1f3f5;color:#495057;"
			};
			const style = styles[data.contract_status] || styles.Draft;
			return `<span style="display:inline-block;padding:2px 10px;border-radius:12px;font-weight:600;${style}">${formatted}</span>`;
		}

		if (column.fieldname === "progress_pct") {
			const pct = Number(data.progress_pct || 0);
			const color = pct >= 90 ? "#2f9e44" : pct >= 50 ? "#f08c00" : "#e03131";
			return `<span style="font-weight:700;color:${color};">${formatted}</span>`;
		}

		if (column.fieldname === "total_outstanding") {
			return `<span style="font-weight:700;color:#e03131;">${formatted}</span>`;
		}

		if (column.fieldname === "total_paid") {
			return `<span style="font-weight:700;color:#2f9e44;">${formatted}</span>`;
		}

		if (column.fieldname === "installment_summary" && String(data.installment_summary || "").includes("overdue")) {
			return `<span style="font-weight:700;color:#c92a2a;">${formatted}</span>`;
		}

		return formatted;
	}
};

File: lms/lms/report/lms_sales_pipeline/lms_sales_pipeline.py
import frappe
from frappe.utils import flt


def execute(filters=None):
	filters = filters or {}
	columns = get_columns()
	data    = get_data(filters)
	summary = get_summary(data)
	chart   = get_chart(data)
	return columns, data, None, chart, summary


def get_columns():
	return [
		{"label": "Contract",           "fieldname": "contract",            "fieldtype": "Link",    "options": "Plot Contract", "width": 150},
		{"label": "Customer",           "fieldname": "customer",            "fieldtype": "Link",    "options": "Customer",      "width": 190},
		{"label": "Plot",               "fieldname": "plot",                "fieldtype": "Link",    "options": "Plot Master",   "width": 130},
		{"label": "Contract Date",      "fieldname": "contract_date",       "fieldtype": "Date",                                "width": 120},
		{"label": "Deadline",           "fieldname": "payment_deadline",    "fieldtype": "Date",                                "width": 120},
		{"label": "Status",             "fieldname": "contract_status",     "fieldtype": "Data",                                "width": 100},
		{"label": "Contract Value (TZS)","fieldname": "selling_price",      "fieldtype": "Float",                               "width": 170},
		{"label": "Paid (TZS)",         "fieldname": "total_paid",          "fieldtype": "Float",                               "width": 150},
		{"label": "Outstanding (TZS)",  "fieldname": "total_outstanding",   "fieldtype": "Float",                               "width": 160},
		{"label": "Progress %",         "fieldname": "progress_pct",        "fieldtype": "Percent",                             "width": 110},
		{"label": "Installments",       "fieldname": "installment_summary", "fieldtype": "Data",                                "width": 140},
	]


def get_data(filters):
	conditions = ["pc.docstatus = 1"]

	if filters.get("contract_status"):
		conditions.append("pc.contract_status = %(contract_status)s")
	else:
		conditions.append("pc.contract_status IN ('Ongoing', 'Completed')")

	if filters.get("customer"):
		conditions.append("pc.customer = %(customer)s")
	if filters.get("from_date"):
		conditions.append("pc.contract_date >= %(from_date)s")
	if filters.get("to_date"):
		conditions.append("pc.contract_date <= %(to_date)s")

	where = " AND ".join(conditions)

	rows = frappe.db.sql(f"""
		SELECT
			pc.name                                                              AS contract,
			pc.customer,
			pc.plot,
			pc.contract_date,
			pc.payment_deadline,
			pc.contract_status,
			pc.selling_price,
			pc.total_paid,
			pc.total_outstanding,
			COUNT(pcp.name)                                                      AS total_inst,
			SUM(CASE WHEN pcp.status = 'Paid'    THEN 1 ELSE 0 END)             AS paid_inst,
			SUM(CASE WHEN pcp.status = 'Overdue' THEN 1 ELSE 0 END)             AS overdue_inst
		FROM `tabPlot Contract` pc
		LEFT JOIN `tabPlot Contract Payment` pcp ON pcp.parent = pc.name
		WHERE {where}
		GROUP BY pc.name
		ORDER BY pc.contract_date DESC
	""", filters, as_dict=True)

	data = []
	for row in rows:
		price = flt(row.selling_price)
		paid  = flt(row.total_paid)
		pct   = (paid / price * 100) if price else 0
		inst_summary = f"{row.paid_inst or 0}/{row.total_inst or 0}"
		if row.overdue_inst:
			inst_summary += f" ({row.overdue_inst} overdue)"
		data.append({
			"contract":             row.contract,
			"customer":             row.customer,
			"plot":                 row.plot,
			"contract_date":        row.contract_date,
			"payment_deadline":     row.payment_deadline,
			"contract_status":      row.contract_status,
			"selling_price":        price,
			"total_paid":           paid,
			"total_outstanding":    flt(row.total_outstanding),
			"progress_pct":         pct,
			"installment_summary":  inst_summary,
		})
	return data


def get_summary(data):
	if not data:
		return []
	total_value       = sum(flt(r["selling_price"])     for r in data)
	total_paid        = sum(flt(r["total_paid"])        for r in data)
	total_outstanding = sum(flt(r["total_outstanding"]) for r in data)
	ongoing_contracts = sum(1 for r in data if r["contract_status"] == "Ongoing")
	completed_contracts = sum(1 for r in data if r["contract_status"] == "Completed")
	progress_pct = (total_paid / total_value * 100) if total_value else 0
	return [
		{"label": "Total Contract Value", "value": total_value,       "datatype": "Currency", "indicator": "Blue"},
		{"label": "Total Paid",           "value": total_paid,        "datatype": "Currency", "indicator": "Green"},
		{"label": "Total Outstanding",    "value": total_outstanding, "datatype": "Currency", "indicator": "Red"},
		{"label": "Ongoing Contracts",    "value": ongoing_contracts, "datatype": "Int", "indicator": "Orange"},
		{"label": "Completed Contracts",  "value": completed_contracts, "datatype": "Int", "indicator": "Green"},
		{"label": "Collection Progress %", "value": progress_pct, "datatype": "Percent", "indicator": "Green" if progress_pct >= 70 else "Orange"},
	]


def get_chart(data):
	if not data:
		return None

	status_order = ["Ongoing", "Completed", "Terminated", "Cancelled", "Draft"]
	status_map = {status: 0 for status in status_order}
	for row in data:
		status = row.get("contract_status")
		if status in status_map:
			status_map[status] += 1

	labels = [status for status in status_order if status_map.get(status)]
	if not labels:
		return None

	color_map = {
		"Ongoing": "#f08c00",
		"Completed": "#2f9e44",
		"Terminated": "#e8590c",
		"Cancelled": "#e03131",
		"Draft": "#868e96",
	}

	return {
		"data": {
			"labels": labels,
			"datasets": [
				{
					"name": "Contracts",
					"values": [status_map[label] for label in labels],
				}
			],
		},
		"type": "donut",
		"colors": [color_map[label] for label in labels],
	}
